[24.1] Drop restriction to switch to immutable histories

davelopez opened the pull request:
Fixes #18113

This allows the user to "set as current" any **owned history** instead of relying on the history not being available in the UI when it is immutable (purged, deleted, or archived) to avoid actions that will result in errors. So now the UI will be more defensive and informative when you switch to one of these histories but still will unlock actions that might be useful like "extract workflow", etc.

## For immutable histories

| The status is displayed in the panel | Menu options are limited |
|--------|-------|
| ![image](https://github.com/galaxyproject/galaxy/assets/46503462/1d899656-2460-4e58-a1ce-09521f37906e)   | ![image](https://github.com/galaxyproject/galaxy/assets/46503462/bd367ea7-726e-4d04-8e1f-2dc193eeb2e1) |


### Tool Run will be disabled
![image](https://github.com/galaxyproject/galaxy/assets/46503462/6378990a-fbba-4187-816f-e5f29879217f)

### Workflow Run will be disabled
| ![Screenshot from 2024-05-27 16-32-23](https://github.com/galaxyproject/galaxy/assets/46503462/c402a1a9-8183-496b-806f-9ffdbf97907b) | ![Screenshot from 2024-05-27 16-32-33](https://github.com/galaxyproject/galaxy/assets/46503462/7a0e5e51-c8c8-4718-947b-2889e323f9a1) |
|--------|-------|

### You can't upload data
![image](https://github.com/galaxyproject/galaxy/assets/46503462/5be0e061-6f7c-4685-8a8d-1338ec8031a0)




## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

davelopez committed:
Drop backend restrictions to switch to immutable histories

davelopez committed:
Move immutability restriction check from HistoryView to HistoryPanel

davelopez committed:
Display history status in Panel

davelopez committed:
Add button to set archived histories as current

davelopez committed:
Drop redundant history state info in HistoryView

The state will be displayed in all history panels

davelopez committed:
Allow to set any owned history as current

davelopez committed:
Disable collection editing for immutable histories

davelopez committed:
Allow to go to storage overview for immutable histories

And handle the controls visibility to avoid history mutation by deleting datasets, etc.

davelopez committed:
Disable unavailable elements in History menu for immutable histories

davelopez committed:
Inform users why some actions for histories might be disabled

davelopez committed:
Avoid running tools on immutable histories

davelopez committed:
Restrict some more options in history menu

mvdbeek commented on client/src/components/Tool/ToolForm.vue:
```diff
@@ -3,6 +3,9 @@
         <b-alert :show="messageShow" :variant="messageVariant">
             {{ messageText }}
         </b-alert>
+        <b-alert v-if="!showLoading && !canMutateHistory" show variant="warning">
```
We probably need the same in the workflow run form, the upload dialog and the collection builders ?

davelopez replied:
I've added similar checks to the workflow run and the upload dialog. Is there a way to reach the collection builders without using the upload dialog?
I hope all cases are covered now, in the worst case it will fail after trying but there is a lot of feedback now on what kind of history we are dealing with.

mvdbeek reviewed (approved):
Looks great to me, however do we maybe want to target 24.1 and put this through a bit of testing ? The thing that worries me just a tiny bit is the change of the server-side current history (in particular it might make some "history-less" future UI options harder, like the workflow-centric UI we've been talking about). This is probably fine, but maybe we could just allow extracting workflows regardless of history status for 24.0? I should've been more specific with 

> Generally speaking I should be able to make a purged history my current history.

that I'd like to be able to have the normal history panel on the right side, but without making it the server-side current history.

davelopez committed:
Avoid workflow run on immutable histories

davelopez committed:
Avoid upload data to immutable histories

davelopez committed:
Support extract workflow from any history in client controller

davelopez committed:
Restore back condition to set current history in backend

The effect of this is that we can switch to any owned history in the UI, but if the history has been deleted it will not change the current history in the backend and a browser refresh will bring back the real current history.

davelopez committed:
Display archived message only to the owner

davelopez committed:
Adapt HistoryView tests

davelopez committed:
Fix conditions for userOwnsHistory and isRegisteredUser

davelopez committed:
Adapt History SelectorModal test
